gordon_janitor_gcp/plugins/authority.py

```diff
@@ -0,0 +1,185 @@
+
+    def _get_or_raise(self, key):
+        item = self.config.get(key)
+        if not item:
+            msg = f'gcp.gce configuration missing "{key}" setting'
+            logging.error(msg)
+            # Close the session if any required config is missing.
+            self.session.close()
+            raise exceptions.GCPConfigError(msg)
+        return item
+
+    def _get_crm_client(self, keyfile_path, scopes):
+        crm_auth = auth.GAuthClient(
+            keyfile_path, scopes=scopes, session=self.session)
+        return gcrm.GCRMClient(crm_auth, self.session)
+
+    def _get_gce_client(self, keyfile_path, scopes):
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        _metadata_blacklist = self.config.get('metadata_blacklist', [])
+        metadata_blacklist = [dict([pair]) for pair in _metadata_blacklist]
+
+        gce_auth = auth.GAuthClient(keyfile_path, scopes=scopes,
+                                    session=self.session)
+        return gce.GCEClient(gce_auth, self.session,
+                             blacklisted_tags=tag_blacklist,
+                             blacklisted_metadata=metadata_blacklist)
+
+    def build_authority(self):
+        """Coordinate dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
+        """
+        self._get_or_raise('zone')
+        keyfile_path = self._get_or_raise('keyfile')
+        scopes = self._get_or_raise('scopes')
+        crm_client = self._get_crm_client(keyfile_path, scopes)
+        gce_client = self._get_gce_client(keyfile_path, scopes)
+
+        return GCEAuthority(self.config, crm_client, gce_client,
+                            self.rrset_channel, **self.kwargs)
+
+
+@zope.interface.implementer(interfaces.IAuthority)
+class GCEAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        crm_client (obj): client used to fetch GCE projects.
+        gce_client (obj): client used to fetch instances for each project.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+
+    def __init__(self, config, crm_client, gce_client, rrset_channel=None,
+                 **kwargs):
+        self.config = config
+        self.crm_client = crm_client
+        self.gce_client = gce_client
+        self.rrset_channel = rrset_channel
+
+    async def _get_active_project_ids(self):
+        active_projects = await self.crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def _get_projects(self):
+        projects = await self._get_active_project_ids()
+        project_blacklist = set(self.config.get('project_blacklist', []))
+        # TODO: emit a metric for all projects and a metric for
+        # project - blacklist.
+        return sorted(projects - project_blacklist)
```

Comment:
OK I'm only have done with my coffee, so pardon me for the silly q: why sorted?

Reply:
Just to introduce determinism into the mix - having the projects always be processed in alphabetical order can help reasoning about/debugging. I can remove it if you think it's confusing?